File: 14. Longest Common Prefix/longestCommonPrefix.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

string longestCommonPrefix(vector<string> &strs) {

  if (strs[0] == "")
    return "";

  string ans = strs[0];

  if (strs.size() > 1) {
    string s;
    for (int i = 1; i < strs.size(); ++i) {
      s = strs[i];
      int j = 0;

      while (ans[j] == s[j] && j < ans.size())
        ++j;

      ans.resize(j);

      if (j == 0)
        return "";
    }
  }
  return ans;
}

/*
       if (strs.size() == 0) return "";
       if (strs.size() == 1) return strs[0];

       sort(strs.begin(), strs.end(), []
         (const string& first, const string& second){
           return first.size() < second.size();
         });

       if (strs[0].empty()) return "";

       string prefix = "";
       int a = 0, i = 0;
       string s = strs[i];
       int sLength = s.size();

       if (isalpha(s[a]) == 0) return "";

       for (int j = 1; j < strs.size(); ++j) {
           string t = strs[j];
           if (t.empty()) return "";

           if (i == 0) {
             if (s[a] != t[a]) return prefix;

             while (s[a] == t[a] && i != j) {
               prefix.push_back(s[a]);
               ++a;
               if (a >= s.size()) break;
             }
             ++i;
           }
           else {
             a = prefix.size() - 1;
             while (prefix[a] != t[a]) {
               prefix.pop_back();
               if (prefix.empty()) return "";
               if (a > 0) {
                   --a;
               }
               else {
                   break;
               }

             }
           }
       }

   return prefix;
   }
*/

int main() { cout << longestCommonPrefix({"flower", "flow", "flight"}); }

/*
["flower","flow","flight"] = "fl"
["dog","racecar","car"] = ""
["",""] = ""
*/

File: 14. Longest Common Prefix/longestCommonPrefix.js
var longestCommonPrefix = function (strs) {
	if (strs.length <= 1 || strs[0] === "") return strs[0] || "";

	/* for loops approach 
  // outer loop iterates through first word 
  for (let i = 0; i < strs[0].length; i++) {
      // inner loop compares letters of first word with each successive word
      for (let j = 1; j < strs.length; j++) {
          if (strs[0][i] != strs[j][i]) { // when match fails, returns slice of first word 
              return strs[0].slice(0,i); // could also use .substr(0,1) as below
          }
      }
  }
  return strs[0]; // returns first word if all words match
  */

	/* every() method approach */
	let i = 0;
	// while all words have the same character at position i, increment i
	while (strs[0][i] && strs.every((word) => word[i] === strs[0][i])) i++;

	// prefix is the substring from the beginning to the last successfully checked i
	return strs[0].substr(0, i); // could also use .slice(0,1) as above
};

console.log(longestCommonPrefix(["flower", "flow", "flight"]));

/*
["flower","flow","flight"] = "fl"
["dog","racecar","car"] = ""
["",""] = ""
*/
